warehouse/manage/views.py

```diff
@@ -1047,6 +1049,119 @@ def toggle_2fa_requirement(self):
         )
 
 
+@view_defaults(
+    context=Project,
+    route_name="manage.project.settings.oidc",
+    renderer="manage/publishing.html",
+    uses_session=True,
+    require_csrf=True,
+    require_methods=False,
+    permission="manage:project",
+    has_translations=True,
+    require_reauth=True,
+    http_cache=0,
+)
+class ManageOIDCProviderViews:
+    def __init__(self, project, request):
+        self.request = request
+        self.project = project
+
+    @property
+    def default_response(self):
+        return {
+            "project": self.project,
+            "github_provider_form": GitHubProviderForm(),
+        }
+
+    @view_config(request_method="GET")
+    def manage_project_oidc_providers(self):
+        return self.default_response
+
+    @view_config(request_method="POST", request_param=GitHubProviderForm.__params__)
+    def add_github_oidc_provider(self):
+        form = GitHubProviderForm(self.request.POST)
+
+        if form.validate():
+            # GitHub OIDC providers are unique on the tuple of
+            # (repository_name, owner, workflow_name), so we check for
+            # an already registered one before creating.
+            provider = (
+                self.request.db.query(GitHubProvider)
+                .filter(
+                    GitHubProvider.repository_name == form.repository,
+                    GitHubProvider.owner == form.owner,
+                    GitHubProvider.workflow_name == form.workflow_name.data,
+                )
+                .one_or_none()
+            )
+            if provider is None:
+                provider = GitHubProvider(
+                    repository_name=form.repository,
+                    owner=form.owner,
+                    owner_id=form.owner_id,
+                    workflow_name=form.workflow_name.data,
+                )
+
+                self.request.db.add(provider)
+
+            # Each project has a unique registration of OIDC providers,
+            # even though the relationship is many-many.
+            if provider in self.project.oidc_providers:
+                self.request.session.flash(
+                    f"{provider} is already registered with {self.project.name}",
+                    queue="error",
+                )
+                return {**self.default_response, "github_provider_form": form}
+
+            self.project.oidc_providers.append(provider)
+
+            self.project.record_event(
+                tag="project:oidc:provider-added",
```

Comment:
Let's make sure to add a special case for these in the template where it gets rendered, so it doesn't show up as the tag name and we can include the `additional` metadata as well.

Additionally, some of the `additional` metadata seems fairly GitHub specific. We should make sure we either differentiate based on providers or publish provider-specific events.

Reply:
> Additionally, some of the additional metadata seems fairly GitHub specific. We should make sure we either differentiate based on providers or publish provider-specific events.

Yeah, this is a little fiddly with our polymorphic model. So far the only common interface between all children of `OIDCProvider` is `__str__`, but we can add more baseline functionality to make the `additional` metadata more useful (and generic).

Reply:
Probably acceptable to just store `str(provider)` here then, and let the class determine how it should be displayed?

Reply:
Yep. sounds good. I'll do the same for the deletion route.